ic-cdk/src/stable.rs

```diff
@@ -0,0 +1,775 @@
+            Self::OutOfBounds => f.write_str("Read exceeds allocated memory"),
+        }
+    }
+}
+
+impl error::Error for StableMemoryError {}
+
+/// A standard implementation of [`StableMemory`].
+///
+/// Useful for creating [`StableWriter`] and [`StableReader`].
+#[derive(Default, Debug, Copy, Clone)]
+pub struct CanisterStableMemory {}
+
+impl StableMemory for CanisterStableMemory {
+    fn stable_size(&self) -> u64 {
+        // SAFETY: ic0.stable64_size is always safe to call.
+        unsafe { ic0::stable64_size() }
+    }
+
+    fn stable_grow(&self, new_pages: u64) -> Result<u64, StableMemoryError> {
+        // SAFETY: ic0.stable64_grow is always safe to call.
+        unsafe {
+            match ic0::stable64_grow(new_pages) {
+                u64::MAX => Err(StableMemoryError::OutOfMemory),
+                x => Ok(x),
+            }
+        }
+    }
+
+    fn stable_write(&self, offset: u64, buf: &[u8]) {
+        // SAFETY: `buf`, being &[u8], is a readable sequence of bytes, and therefore valid to pass to ic0.stable64_write.
+        unsafe {
+            ic0::stable64_write(offset, buf.as_ptr() as u64, buf.len() as u64);
+        }
+    }
+
+    fn stable_read(&self, offset: u64, buf: &mut [u8]) {
+        // SAFETY: `buf`, being &mut [u8], is a writable sequence of bytes, and therefore valid to pass to ic0.stable64_read.
+        unsafe {
+            ic0::stable64_read(buf.as_ptr() as u64, offset, buf.len() as u64);
+        }
+    }
+}
+
+/// Gets current size of the stable memory (in WASM pages).
+pub fn stable_size() -> u64 {
+    CANISTER_STABLE_MEMORY.stable_size()
+}
+
+/// Attempts to grow the stable memory by `new_pages` (added pages).
+///
+/// Returns an error if it wasn't possible. Otherwise, returns the previous
+/// size that was reserved.
+///
+/// *Note*: Pages are 64KiB in WASM.
+pub fn stable_grow(new_pages: u64) -> Result<u64, StableMemoryError> {
+    CANISTER_STABLE_MEMORY.stable_grow(new_pages)
+}
+
+/// Writes data to the stable memory location specified by an offset.
+///
+/// Warning - this will panic if `offset + buf.len()` exceeds the current size of stable memory.
+/// Use `stable_grow` to request more stable memory if needed.
+pub fn stable_write(offset: u64, buf: &[u8]) {
+    CANISTER_STABLE_MEMORY.stable_write(offset, buf)
+}
+
+/// Reads data from the stable memory location specified by an offset.
+pub fn stable_read(offset: u64, buf: &mut [u8]) {
+    CANISTER_STABLE_MEMORY.stable_read(offset, buf)
+}
+
+/// Returns a copy of the stable memory.
+///
+/// This will map the whole memory (even if not all of it has been written to).
+///
+/// # Panics
+///
+/// When the bytes of the stable memory cannot fit into a `Vec` which constrained by the usize.
+pub fn stable_bytes() -> Vec<u8> {
+    let size = (stable_size() << 16)
+        .try_into()
+        .expect("overflow: stable memory too large to read in one go");
+    let mut vec = Vec::with_capacity(size);
```

Comment:
Could we use Result rather than panic in the interface here? The `.expect(...)` is up to us, and we could use Vec::try_with_capacity().

Reply:
The content of this `src/stable.rs` file was directly copied from `src/api/stable/mod.rs`.
I will improve the internal implementation in a later PR.